gips/0053-enabling-substreams-based-subgraphs.md

```diff
@@ -0,0 +1,138 @@
+  { label = "substreams-provider-mainnet",
+    details = { type = "substreams",
+    url = "https://mainnet-substreams-url.grpc.substreams.io/",
+    token = "exampletokenhere" }},
+]
+```
+Substreams may require dedicated error handling / retry logic, though a lot of the requirements should be covered by the Firehose cursor-based integration.
+
+#### Mapping entities
+Substreams have their own protobuf schemas. However to directly copy the raw substream schema would be to lose the ability to link and traverse entities in the resulting GraphQL schema.
+
+Utilities should certainly be provided to auto-generate a `.graphql` file from a substream protobuf schema, with intelligent type mapping. However once generated, developers should be able to update the `.graphql` schema, to identify connections between entities, and to add derived fields. Therefore there is still value in having a schema file as part of the subgraph definition.
+
+This proposal does assume a tight coupling between entity names in the substream, and entity names in the subgraph
+
+#### Updating entities
+In order to support performant querying, Graph Node stores full entities at every block where those entities change. It supports “upsert” type behaviour, where saved values are merged with existing entities.
+
+Substreams currently provide entity “deltas”, which just capture the changed values. Therefore there is a requirement to fetch any prior entity values to create the full entity to store. This could be done on either the Substream or Graph Node side of the integration, so we should work through the trade offs & alternatives.
+
+Graph Node will still need to “close” the block range for any existing entities, which could negatively impact indexing performance.
+
+#### Deterministic indexing
+A fundamental requirement is deterministic indexing, specifically the generation of a Proof of Indexing, which can then be cross-checked across indexers.
+
+> Graph Node makes the assumption that all upstream datasources are deterministic, and will make the same assumption about Substreams
+
+Depending on the implementation in Graph Node, this could leverage the existing Proof of Indexing setup, or it may need a dedicated Proof of Indexing in Graph Node.
+
+#### Handling re-orgs
+Graph Node will still need to be re-org aware. The substreams connection should provide the relevant information (which blocks to remove). Graph Node will need to remove entities from those blocks, and re-open the block range for older entities. This functionality already exists in Graph Node, but might need to be updated for the new data source.
+
+#### Linear process & batch back-filling
+Graph node will need to support linear processing, e.g. when indexing at the chain head. There is a possibility that linear processing historical blocks becomes a significant performance bottleneck. In that case, a dedicated “batch” insert implementation might be required to load historical data into the subgraph.
```

Comment:
The latter "batch" implementation exists, but it is a side car and not the focus of the implementation

Reply:
Revised to read "In that case, a dedicated “batch” insert implementation exists, but it is not the focus of the implementation."